ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,858 @@
+  std::string prevParentName = "";
+
+  // Store the ogre objects in a vecotr
+  std::vector<Ogre::MovableObject *> ogreObjects;
+  while (itor.hasMoreElements())
+  {
+    Ogre::MovableObject *object = itor.peekNext();
+    ogreObjects.push_back(object);
+    itor.moveNext();
+  }
+
+  // Sort the ogre objects by name
+  // The algorithm of handeling multi-link models depends on a sorted objects
+  // by name, so all links that belongs to the same object come in order
+  std::sort(ogreObjects.begin(), ogreObjects.end(),
+    [] (Ogre::MovableObject * object1, Ogre::MovableObject * object2) {
+      return object1->getName() > object2->getName();
+  });
+
+  for (auto object : ogreObjects)
+  {
+    Ogre::Item *item = static_cast<Ogre::Item *>(object);
+
+    // get visual from ogre item
+    Ogre::Any userAny = item->getUserObjectBindings().getUserAny();
+
+    if (!userAny.isEmpty() && userAny.getType() == typeid(unsigned int))
+    {
+      // get visual id for the ogre item
+      auto visualId = Ogre::any_cast<unsigned int>(userAny);
+
+      VisualPtr visual;
+      try
+      {
+        visual = this->scene->VisualById(visualId);
+      }
+      catch(Ogre::Exception &e)
+      {
+        ignerr << "Ogre Error:" << e.getFullDescription() << "\n";
+      }
+      Ogre2VisualPtr ogreVisual = std::dynamic_pointer_cast<Ogre2Visual>(
+        visual);
+
+      // get class user data
+      Variant labelAny = ogreVisual->UserData(this->labelKey);
+      int label;
+      try
+      {
+        label = std::get<int>(labelAny);
+      }
+      catch(std::bad_variant_access &e)
+      {
+        // items with no class are considered background
+        label = this->backgroundLabel;
+      }
+
+      // sub item custom parameter to set the pixel color material
+      Ogre::Vector4 customParameter;
+
+      // Material Switching
+      if (this->type == SegmentationType::SEMANTIC)
+      {
+        if (this->isColoredMap)
+        {
+          // semantic material(each pixel has item's color)
+          math::Color color = this->LabelToColor(label);
+          customParameter = Ogre::Vector4(
+            color.R(), color.G(), color.B(), 1.0);
+        }
+        else
+        {
+          // labels ids material(each pixel has item's label)
+          float labelColor = label / 255.0;
+          customParameter = Ogre::Vector4(
+            labelColor, labelColor, labelColor, 1.0);
+        }
+      }
+      else if (this->type == SegmentationType::PANOPTIC)
+      {
+        auto itemName = visual->Name();
+        std::string parentName;
+
+        // Visuals that are added without a name (through rendering API)
+        // will have the name "scene::Visual(VISUAL_ID)", use it as a name
+        auto unNamedVisualIndex = itemName.find("scene::Visual");
+        if (unNamedVisualIndex != std::string::npos)
+          parentName = itemName.substr(0, itemName.find_first_of(")") + 1);
```

Comment:
instead of searching for a hardcoded name string, how about checking to see if the visual shares the same parent as the previous visual, something like:

```
auto parent = visual->Parent();
if (parent && parent == prevParent)
  isMultiLink = true;
```

Reply:
@iche033 
mmm, but It is not the direct parent in all cases, I check if all the visuals has the same top parent, and it might be not the direct parent

for example, if we have a model which contains many links, so the visuals names will be something like:
`model::link1::visual1` & `model::link2::visual2` , so I need to check the top parent which is `model` but not the direct parent which is `link1` or `link2`

Reply:
I see, how about adding a convenient function to get the top level model visual?

```cpp
////////////////////////////////////////////////
VisualPtr Ogre2SegmentationMaterialSwitcher::TopLevelVisual(VisualPtr _visual)
{ 
  if (!_visual)
    return _visual;
  VisualPtr p = _visual;
  while (p->Parent() && p->Parent() != _visual->Scene()->RootVisual())
    p = std::dynamic_pointer_cast<Visual>(p->Parent());
  return p;
}
```

then you can get the parent name here by calling 

```cpp
   parentName = this->TopLevelVisual(visual)->Name();
```

this is just to prevent relying on hardcoded name prefixes in case the visual naming style changes in the future.

Reply:
@iche033 I went ahead and took your `TopLevelModelVisual` approach in https://github.com/ignitionrobotics/ign-rendering/pull/329/commits/4f0209d5697d7b58684a0772568c2e916fd1179f, and it seems to work well for models with multiple links. So, I'll mark this as resolved